v0/destinations/airship/transform.js

```diff
@@ -0,0 +1,295 @@
+  if (timestamp) response.body.JSON.occured = timestamp;
+  response.body.JSON.user = {};
+  response.body.JSON.user.named_user_id = message.userId;
+  response.headers = {
+    "Content-Type": "application/json",
+    Accept: "application/vnd.urbanairship+json; version=3",
+    "X-UA-Appkey": `${appKey}`,
+    Authorization: `Bearer ${apiKey}`
+  };
+  if (message.type === "track") {
+    response.endpoint = `${BASE_URL}/api/custom-events`;
+  }
+  response.method = defaultPostRequestConfig.requestMethod;
+  response.body.JSON.body = {};
+  response.body.JSON.body = payload;
+  return response;
+};
+
+const groupResponseBuilder = (message, { Config }) => {
+  const tagPayload = constructPayload(message, groupMapping);
+  const attributePayload = constructPayload(message, groupMapping);
+  const { appKey, dataCenter, appSecret } = Config;
+  if (!appKey || !appSecret) {
+    if (!appKey)
+      throw new CustomError(
+        "[Airship] App Key is required for Authentication",
+        400
+      );
+    else
+      throw new CustomError(
+        "[Airship] App Secet is required for authentication",
+        400
+      );
+  }
+
+  let BASE_URL = BASE_URL_US;
+  // check the region and which api end point should be used
+  BASE_URL = dataCenter ? BASE_URL_EU : BASE_URL;
+
+  const traits = flattenJson(getFieldValueFromMessage(message, "traits"));
+  if (!isDefinedAndNotNull(traits)) {
+    throw new CustomError(
+      "[Airship]:: for identify, tags or attributes properties are required under traits",
+      400
+    );
+  }
+
+  tagPayload.add = { rudderstack_integration: [] };
+  tagPayload.remove = { rudderstack_integration: [] };
+  const timestamp = getFieldValueFromMessage(message, "timestamp");
+  attributePayload.attributes = [];
+  Object.keys(traits).forEach(key => {
+    if (typeof traits[key] === "boolean") {
+      const tag = key.toLowerCase().replace(/\./g, "_");
+      if (traits[key] === true) {
+        tagPayload.add.rudderstack_integration_group.push(tag);
+      }
+      if (traits[key] === false) {
+        tagPayload.remove.rudderstack_integration_group.push(tag);
+      }
+    }
+    if (typeof traits[key] !== "boolean") {
+      const attribute = {};
+      attribute.action = "set";
+      const keyMapped = RESERVED_TRAITS_MAPPING[key.toLowerCase()];
+      if (keyMapped) {
+        attribute.key = keyMapped;
+      } else {
+        attribute.key = key.replace(/\./g, "_");
+      }
+      attribute.value = traits[key];
+      attribute.timestamp = timestamp;
+      attributePayload.attributes.push(attribute);
+    }
+  });
+
+  let tagResponse;
+  let attributeResponse;
+  if (
+    tagPayload.add.rudderstack_integration_group.length ||
+    tagPayload.remove.rudderstack_integration_group.length
+  ) {
+    if (!tagPayload.add.rudderstack_integration_group.length)
+      delete tagPayload.add;
+    if (!tagPayload.remove.rudderstack_integration_group.length)
+      delete tagPayload.remove;
```

Comment:
Do we need this check? If its empty airship should not be throwing errors anyway? This is making the code messy

Reply:
Resolved